### PR TITLE
Process commitments in a separate thread

### PR DESCRIPTION
**Pull request by kpp**
# Description

>and one major problem was that we were late to send btc to the sequencer, since the txs failed the commitments all failedfinally we did a huge commitment for l1 block 1000 to l1 block 6000
in real scenarios that would be really hard to prove and fit in a bitcoin block

## Linked Issues
- Fixes https://github.com/chainwayxyz/citrea/issues/574

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

### CODE REVIEW

File: crates/sequencer/src/commitment_controller.rs [hunk 1]
```diff
@@ -46,15 +46,31 @@ pub fn get_commitment_info(
...
+            // Take while sum of l2 ranges <= min_soft_confirmations_per_commitment
+            while l2_length <= min_soft_confirmations_per_commitment
+                || l1_start < prev_l1_height - 1
+            {
+                let Some((l2_start, l2_end)) =
+                    ledger_db.get_l2_range_by_l1_height(SlotNumber(l1_end))?
+                else {
+                    bail!("Sequencer: Failed to get L1 L2 connection");
+                };
+                let l2_range_length = l2_end.0 - l2_start.0;
```
Comment (eyusufatik) [inline crates/sequencer/src/commitment_controller.rs:61]: l2_range_length = l2_end.0 - l2_start.0 + 1 ;

both is inclusive there

File: crates/sequencer/src/commitment_controller.rs [hunk 2]
```diff
@@ -46,15 +46,31 @@ pub fn get_commitment_info(
     // if there is a height then start from height + 1 and go to prev_l1_height
     let (l2_range_to_submit, l1_height_range) = match last_commitment_l1_height {
         Some(last_commitment_l1_height) => {
-            let l1_height_range = (last_commitment_l1_height.0 + 1, prev_l1_height);
+            let l1_start = last_commitment_l1_height.0;
+            let mut l1_end = l1_start;
+            let mut l2_length = 0;
+            // Take while sum of l2 ranges <= min_soft_confirmations_per_commitment
+            while l2_length <= min_soft_confirmations_per_commitment
```
Comment (eyusufatik) [inline crates/sequencer/src/commitment_controller.rs:53]: l2_length < min_soft_confirmations_per_commitment


we can commit on l2_length = min_soft_confirmations_per_commitment

File: crates/sequencer/src/commitment_controller.rs [hunk 3]
```diff
@@ -46,15 +46,31 @@ pub fn get_commitment_info(
...
+                else {
+                    bail!("Sequencer: Failed to get L1 L2 connection");
+                };
+                let l2_range_length = l2_end.0 - l2_start.0;
+                l2_length += l2_range_length;
+                l1_end += 1;
+            }
+            let l1_height_range = (l1_start + 1, l1_end);
 
             let Some((l2_start_height, _)) =
```
Comment (eyusufatik) [inline crates/sequencer/src/commitment_controller.rs:67]: you're already querying those connections above, just keep tract of l2 start and l2 end :D

File: crates/sequencer/src/commitment_controller.rs [hunk 4]
```diff
@@ -46,15 +46,31 @@ pub fn get_commitment_info(
     // if there is a height then start from height + 1 and go to prev_l1_height
     let (l2_range_to_submit, l1_height_range) = match last_commitment_l1_height {
         Some(last_commitment_l1_height) => {
-            let l1_height_range = (last_commitment_l1_height.0 + 1, prev_l1_height);
+            let l1_start = last_commitment_l1_height.0;
```
Comment (eyusufatik) [inline crates/sequencer/src/commitment_controller.rs:49]: set l1_start = last_commitment_l1_height.0 + 1

for easier understandability in the future, also below works incorrectly bc of this :D

Comment (eyusufatik) [reply crates/sequencer/src/commitment_controller.rs:49]: then line 65

let l1_height_range = (l1_start + 1, l1_end);


becomes

let l1_height_range = (l1_start, l1_end);

File: crates/sequencer/src/sequencer.rs [hunk 5]
```diff
@@ -591,6 +542,100 @@ where
...
+                        .set_last_sequencer_commitment_l1_height(SlotNumber(
+                            commitment_info.l1_height_range.end().0,
+                        ))
+                        .expect("Sequencer: Failed to set last sequencer commitment L1 height");
+
+                    warn!("Commitment info: {:?}", commitment_info);
+                    let l1_start_height = commitment_info.l1_height_range.start().0;
+                    let l1_end_height = commitment_info.l1_height_range.end().0;
+                    let l2_start = l2_range_to_submit.start().0 as u32;
+                    let l2_end = (l2_range_to_submit.end().0 + 1) as u32;
```
Comment (eyusufatik) [inline crates/sequencer/src/sequencer.rs:609]: there should be no +1 

@ercecan 😠

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/sequencer/src/commitment_controller.rs (hunk 2 above)
Comment (kpp) [reply crates/sequencer/src/commitment_controller.rs:53]: I dont understand

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**